crates/wasm-pkg-client/src/caching/mod.rs

```diff
@@ -0,0 +1,80 @@
+use std::future::Future;
+use std::ops::Deref;
+
+use bytes::Bytes;
+use futures_util::stream::BoxStream;
+use wasm_pkg_common::{digest::ContentDigest, package::PackageRef, Error};
+
+use crate::{Client, Release};
+
+mod file;
+
+pub use file::FileCache;
+
+/// A trait for a cache of data.
+pub trait Cache {
+    /// Puts the data with the given hash into the cache
+    fn put_data(
+        &self,
+        digest: ContentDigest,
+        data: BoxStream<Result<Bytes, Error>>,
+    ) -> impl Future<Output = Result<(), Error>> + Send;
+    /// Gets the data with the given hash from the cache. Returns None if the data is not in the cache.
+    fn get_data(
+        &self,
+        digest: &ContentDigest,
+    ) -> impl Future<Output = Result<Option<BoxStream<Result<Bytes, Error>>>, Error>> + Send;
+}
+
+pub struct CachingClient<T> {
+    pub client: Client,
+    cache: T,
+}
+
+impl<T> AsRef<Client> for CachingClient<T> {
+    fn as_ref(&self) -> &Client {
+        &self.client
+    }
+}
+
+impl<T> Deref for CachingClient<T> {
```

Comment:
Using Deref this way is generally discouraged: https://rust-unofficial.github.io/patterns/anti_patterns/deref.html

Reply:
I can go either way on this. From that link:

> There is no one good alternative.

Which makes me laugh because if it is an antipattern, then hopefully they'd have something else you can do 😆 

The purpose here was not to have polymorphism so much as a thin wrapper around the normal client to avoid the need of doing something like `client.client.method` since you'll still need to do things like fetch the `Release`. 

I am not married to this though. If you're strongly opposed it is easy enough to drop. Just let me know!

Reply:
For me, the important part is "this is a surprising idiom"; I was indeed confused at first about how this was supposed to work. `Deref` is _mostly_ used only for "smart pointer"-style generic wrappers; things that add actual functionality more often expose inner data with `AsRef` or full delegation.

Reply:
Ok, I removed it